autodocodec/src/Autodocodec/Codec.hs

```diff
@@ -156,6 +157,30 @@ data Codec context input output where
     -- | Codec for the 'Right' side
     Codec context input2 output2 ->
     Codec context (Either input1 input2) (Either output1 output2)
+  -- | Encode/decode a discriminated union of objects
+  --
+  -- Type type of object being encoded/decoded is discriminated by
+  -- a designated "discriminator" property on the object which takes a string value.
+  --
+  -- When encoding, the input the provided function is applied to obtain the
+  -- discriminator value and a value of an existentially-quantified type 'SomeEncodable'
+  -- which is used to encode the object.
+  --
+  -- When decoding, the value of the discriminator property is looked up in the `InsOrdHashMap`
+  -- to obtain a value of an existentially-quantified type 'SomeDecodeable output' which
+  -- can be used to decode the output.
+  --
+  -- The 'InsOrdHashMap' is also used to generate schemas for the type.
+  -- In particular, for OpenAPI 3, it will generate a schema with a 'discriminator', as defined
+  -- by https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/
+  DiscriminatedUnionCodec ::
+    -- | propertyName to use for discrimination
+    Text ->
+    -- | how to encode the input
+    (input -> (Discriminator, SomeEncodable)) ->
+    -- | how to decode the output
+    InsOrdHashMap Discriminator (SomeDecodable output) ->
```

Comment:
We definitely don't want to have `autodocodec` to have to depend on `insert-ordered-containers` because it has an ENORMOUS dependency footprint (`lens`).

Reply:
No problem. We can use a normal `HashMap`. The only reason I used `InsOrdHashMap` is that that's what `openapi3` uses.

Reply:
Done.